include/boost/graph/isomorphism.hpp

```diff
@@ -158,46 +161,63 @@ namespace detail
             );
         }
 
+        // Generates map of invariant multiplicity from sorted invariants
+        template<typename ForwardIterator>
+        multiplicity_map multiplicities(ForwardIterator first, const ForwardIterator last) 
+        {
+            typedef typename multiplicity_map::iterator invar_map_iter;
+
+            assert(std::is_sorted(first, last));
+            multiplicity_map invar_multiplicity;
+
+            if(first == last) 
+                return invar_multiplicity;
+
+            invariant_t invar = *first;
+            invar_map_iter inserted = invar_multiplicity.emplace(invar, 1).first;
+            ++first;
+            for(; first != last; ++first)
+            {
+                if(*first == invar)
+                {
+                    inserted->second += 1;
+                }
+                else
+                {
+                    invar = *first;
+                    inserted = invar_multiplicity.emplace(invar, 1).first;
+                }
+            }
+
+            return invar_multiplicity;
+        }
+
         bool test_isomorphism()
         {
             // reset isomapping
             BGL_FORALL_VERTICES_T(v, G1, Graph1)
             f[v] = graph_traits< Graph2 >::null_vertex();
 
-            std::size_t max_invariant = 0;
-            {
-                std::vector< invar1_value > invar1_array;
-                invar1_array.reserve(num_vertices(G1));
-                BGL_FORALL_VERTICES_T(v, G1, Graph1)
-                  invar1_array.push_back(invariant1(v));
-                sort(invar1_array);
-
-                std::vector< invar2_value > invar2_array;
-                invar2_array.reserve(num_vertices(G2));
-                BGL_FORALL_VERTICES_T(v, G2, Graph2)
-                  invar2_array.push_back(invariant2(v));
-                sort(invar2_array);
-                if (!equal(invar1_array, invar2_array))
-                    return false;
-
-                // Empty graphs case is covered before test_isomorphism is
-                // called, so the invar?_arrays cannot be empty, so back() is
-                // safe. Also the two invariant arrays are equal:
-                max_invariant = invar1_array.back();
-                assert(max_invariant != std::numeric_limits<std::size_t>::max());
-                max_invariant += 1;
-            }
-
+            // Calculate all invariants of G1 and G2, sort and compare
+            std::vector< invariant_t > invar1_array;
+            invar1_array.reserve(num_vertices(G1));
+            BGL_FORALL_VERTICES_T(v, G1, Graph1)
+            invar1_array.push_back(invariant1(v));
+            sort(invar1_array);
+
+            std::vector< invariant_t > invar2_array;
+            invar2_array.reserve(num_vertices(G2));
+            BGL_FORALL_VERTICES_T(v, G2, Graph2)
+            invar2_array.push_back(invariant2(v));
+            sort(invar2_array);
+            if (!equal(invar1_array, invar2_array))
+                return false;
```

Comment:
Are the number of vertices in G1 and G2 guaranteed to be equal here? 

If not, then one minor optimization we can do is to check that the number of vertices in both graphs is equal before going ahead with creating the arrays and sorting them, etc.

Reply:
Yes, this is guaranteed by [an earlier check](https://github.com/boostorg/graph/blob/b737cca58ae6d31cdd4241ff4de7b4755ba614a7/include/boost/graph/isomorphism.hpp#L596).